Support for Set Language Separately (#929)
smartdevicelink/sdl_evolution: proposals/NNNN-set-language-separately.md

> # Set language separately

Reviewer: This should be updated to `Support fo
```suggestion
# Support for Set Language Separately
```

Please make the title of your pull request the exact title of your proposal.

* Proposal: [SDL-NNNN](NNNN-set-language-separately.md)

> * Author: [zhouxin627](https://github.com/zhouxin627)

Reviewer: Can you please include your name here?

Author: @theresalech Thanks, fixed in [7e8d6ec](https://github.com/smartdevicelink/sdl_evolution/pull/929/commits/7e8d6ec4da4bcbf1ef2512c946d8870da7792adc)

* Status: **Awaiting review**
* Impacted Platforms: [iOS / Java Suite]

## Introduction
This proposal provides support for updating to two separate languages.

## Motivation
According to RPC Spec documents, it seems like when the HU's `hmiDisplayLanguage` or `VR/TTS` language is different from the app's desired language, the app will send `ChangeRegistration` to change the registered language which is supported.

> But at the SDL Proxy sourcecode's side, only the difference of `VR/TTS` language will trigger `ChangeRegistration`.

Reviewer: ```suggestion
But at the SDL Proxy source code's side, only the difference of `VR/TTS` language will trigger `ChangeRegistration`.

```

The Proxy does not currently support updating to two separate languages.
In other words, the developer can only be told that the language updated, but not which language type.
The `hmiDisplayLanguage` and `VR/TTS` language could be set separately in some use cases for OEM.

## Proposed solution
The proposed solution is to add a new method to pass the HMI display language separately, and the old one should be deprecated.

### iOS Library
A new SDLManagerDelegate method would need to be added, and the old one should be deprecated:

```objc

> 	 * Called when the lifecycle manager detected a language mismatch. In case of a language mismatch the manager should change the apps registration by updating the lifecycle configuration to the specified language. If the app can support the specified language it should return an Object of SDLLifecycleConfigurationUpdate, otherwise it should return nil to indicate that the language is not supported.

Reviewer: ```suggestion
	 * Called when the lifecycle manager detected a language mismatch. In case of a language mismatch the manager should change the app's registration by updating the lifecycle configuration to the specified language. If the app can support the specified language it should return an Object of SDLLifecycleConfigurationUpdate, otherwise it should return nil to indicate that the language is not supported.

```

	 *

> 	 * @param language The language of the connected head unit the manager is trying to update the configuration.

Reviewer: ```suggestion
	 * @param language The language of the connected head unit for which the manager is trying to update the configuration.

```

> +	 * @param hmiLanguage The language of the connected head unit the manager is trying to update the configuration.

Reviewer: ```suggestion
+	 * @param hmiLanguage The language of the connected head unit for which the manager is trying to update the configuration.

```

	 * @return An object of SDLLifecycleConfigurationUpdate if the head unit language is supported, otherwise nil to indicate that the language is not supported.
	 */
-	- (nullable SDLLifecycleConfigurationUpdate *)managerShouldUpdateLifecycleToLanguage:(SDLLanguage)language;

> +	- (nullable SDLLifecycleConfigurationUpdate *)managerShouldUpdateLifecycleToLanguage:(SDLLanguage)language hmiLanguage:(SDLLanguage)hmiLanguage;

Reviewer: The extra `+`s and `-`s here should be removed.  Please add the actual code for deprecating the method, or at least include a comment like "deprecate this method".  Lastly, please move the code to the left so that it starts at the far left column.

Author: @theresalech Thanks, fixed in [7e8d6ec](https://github.com/smartdevicelink/sdl_evolution/pull/929/commits/7e8d6ec4da4bcbf1ef2512c946d8870da7792adc)

```

### Java Library
A new SdlManagerListener method would need to be added, and the old one should be deprecated:
```java

> 	 * @param language The language of the connected head unit the manager is trying to update the configuration.

Reviewer: ```suggestion
	 * @param language The language of the connected head unit for which the manager is trying to update the configuration.

```

> +	 * @param hmiLanguage The hmiLanguage of the connected head unit the manager is trying to update the configuration.

Reviewer: ```suggestion
+	 * @param hmiLanguage The hmiLanguage of the connected head unit for which the manager is trying to update the configuration.

```

	 * @return An object of LifecycleConfigurationUpdate if the head unit language is supported,
	 * otherwise null to indicate that the language is not supported.
	 */
-	LifecycleConfigurationUpdate managerShouldUpdateLifecycle(Language language);

> +	LifecycleConfigurationUpdate managerShouldUpdateLifecycle(Language language, Language hmiLanguage);

Reviewer: The extra `+`s and `-`s here should be removed.  Please add the actual code for deprecating the method, or at least include a comment like "deprecate this method".  Lastly, please move the code to the left so that it starts at the far left column.

Author: @theresalech Thanks, fixed in [7e8d6ec](https://github.com/smartdevicelink/sdl_evolution/pull/929/commits/7e8d6ec4da4bcbf1ef2512c946d8870da7792adc)

But at the SDL Proxy source code's side, only the difference of `VR/TTS` language will trigger `ChangeRegistration`.

The Proxy does not currently support updating to two separate languages.
In other words, the developer can only be told that the language updated, but not which language type.
The `hmiDisplayLanguage` and `VR/TTS` language could be set separately in some use cases for OEM.

## Proposed solution
The proposed solution is to add a new method to pass the HMI display language separately, and the old one should be deprecated.

### iOS Library
A new SDLManagerDelegate method would need to be added, and the old one should be deprecated:

```objc
* Called when the lifecycle manager detected a language mismatch. In case of a language mismatch the manager should change the app's registration by updating the lifecycle configuration to the specified language. If the app can support the specified language it should return an Object of SDLLifecycleConfigurationUpdate, otherwise it should return nil to indicate that the language is not supported.

*

> * @param language The language of the connected head unit for which the manager is trying to update the configuration.

Reviewer: ```suggestion
/**
* @param language The language of the connected head unit for which the manager is trying to update the configuration.
```
Adding the opening tag of the comment should fix the syntax highlighting.

* @param hmiLanguage The language of the connected head unit for which the manager is trying to update the configuration.

* @return An object of SDLLifecycleConfigurationUpdate if the head unit language is supported, otherwise nil to indicate that the language is not supported.
*/
- (nullable SDLLifecycleConfigurationUpdate *)managerShouldUpdateLifecycleToLanguage:(SDLLanguage)language hmiLanguage:(SDLLanguage)hmiLanguage;

```

### Java Library
A new SdlManagerListener method would need to be added, and the old one should be deprecated:
```java

> * @param language The language of the connected head unit for which the manager is trying to update the configuration.

Reviewer: ```suggestion
/**
* @param language The language of the connected head unit for which the manager is trying to update the configuration.
```
Adding the opening tag of the comment should fix the syntax highlighting.

* Proposal: [SDL-NNNN](NNNN-set-language-separately.md)
* Author: [Zhou Xin](https://github.com/zhouxin627)
* Status: **Awaiting review**
* Impacted Platforms: [iOS / Java Suite]

## Introduction
This proposal provides support for updating to two separate languages.

## Motivation
According to RPC Spec documents, it seems like when the HU's `hmiDisplayLanguage` or `VR/TTS` language is different from the app's desired language, the app will send `ChangeRegistration` to change the registered language which is supported.
But at the SDL Proxy source code's side, only the difference of `VR/TTS` language will trigger `ChangeRegistration`.

The Proxy does not currently support updating to two separate languages.
In other words, the developer can only be told that the language updated, but not which language type.
The `hmiDisplayLanguage` and `VR/TTS` language could be set separately in some use cases for OEM.

## Proposed solution
The proposed solution is to add a new method to pass the HMI display language separately, and the old one should be deprecated.

### iOS Library
A new SDLManagerDelegate method would need to be added, and the old one should be deprecated:

```objc

> * Called when the lifecycle manager detected a language mismatch. In case of a language mismatch the manager should change the app's registration by updating the lifecycle configuration to the specified language. If the app can support the specified language it should return an Object of SDLLifecycleConfigurationUpdate, otherwise it should return nil to indicate that the language is not supported.

Reviewer: ```suggestion
/**
* Called when the lifecycle manager detected a language mismatch. In case of a language mismatch the manager should change the app's registration by updating the lifecycle configuration to the specified language. If the app can support the specified language it should return an Object of SDLLifecycleConfigurationUpdate, otherwise it should return nil to indicate that the language is not supported.
```

*

> /**

Reviewer: ```suggestion
```

* @param language The language of the connected head unit for which the manager is trying to update the configuration.
* @param hmiLanguage The language of the connected head unit for which the manager is trying to update the configuration.

* @return An object of SDLLifecycleConfigurationUpdate if the head unit language is supported, otherwise nil to indicate that the language is not supported.
*/

> - (nullable SDLLifecycleConfigurationUpdate *)managerShouldUpdateLifecycleToLanguage:(SDLLanguage)language hmiLanguage:(SDLLanguage)hmiLanguage;

Reviewer: ```suggestion
- (nullable SDLLifecycleConfigurationUpdate *)manager shouldUpdateLifecycleToLanguage:(SDLLanguage)language hmiLanguage:(SDLLanguage)hmiLanguage;

```
